File: ruby/support/app_identity/support.rb
# frozen_string_literal: true

require "securerandom"
require "digest/sha2"

# This module should only be used by the unit tests and the test suite
# generator.
module AppIdentity::Support # :nodoc:
  private

  def make_app(version, fuzz = nil)
    case version
    when 1 then AppIdentity::App.new(v1_input(fuzz))
    when 2 then AppIdentity::App.new(v2_input(fuzz))
    when 3 then AppIdentity::App.new(v3_input(fuzz))
    when 4 then AppIdentity::App.new(v4_input(fuzz))
    end
  end

  def v1_input(fuzz = nil)
    {version: 1, id: SecureRandom.uuid, secret: SecureRandom.hex(32)}.tap { |app|
      app[:config] = {fuzz: fuzz} if fuzz
    }
  end

  def v2_input(fuzz = nil)
    v1_input(fuzz).merge(version: 2)
  end

  def v3_input(fuzz = nil)
    v1_input(fuzz).merge(version: 3)
  end

  def v4_input(fuzz = nil)
    v1_input(fuzz).merge(version: 4)
  end

  def v1(fuzz = nil)
    @v1 ||= v1_input(fuzz)
  end

  def v1_app(fuzz = nil)
    AppIdentity::App.new(v1(fuzz))
  end

  def v2(fuzz = nil)
    @v2 ||= v2_input(fuzz)
  end

  def v2_app(fuzz = nil)
    AppIdentity::App.new(v2)
  end

  def v3(fuzz = nil)
    @v3 ||= v3_input(fuzz)
  end

  def v3_app(fuzz = nil)
    AppIdentity::App.new(v3)
  end

  def v4(fuzz = nil)
    @v4 ||= v4_input(fuzz)
  end

  def v4_app(fuzz = nil)
    AppIdentity::App.new(v4)
  end

  def build_padlock(app, opts = {})
    app_id = opts.delete(:id) { app.id }
    nonce = opts.delete(:nonce) { "nonce" }
    secret = opts.delete(:secret) { app.secret }
    secret = secret.call if secret.respond_to?(:call)
    version = opts.delete(:version) { app.version }

    case version
    when 1, 2
      Digest::SHA256.hexdigest([app_id, nonce, secret].join(":")).upcase
    when 3
      Digest::SHA384.hexdigest([app_id, nonce, secret].join(":")).upcase
    when 4
      Digest::SHA512.hexdigest([app_id, nonce, secret].join(":")).upcase
    end
  end

  def build_proof(app, padlock, opts = {})
    app_id = opts.delete(:id) { app.id }
    nonce = opts.delete(:nonce) { "nonce" }
    version = opts.delete(:version) { 1 }

    proof = version == 1 ? "#{app_id}:#{nonce}:#{padlock}" : "#{version}:#{app_id}:#{nonce}:#{padlock}"

    Base64.urlsafe_encode64(proof)
  end

  def decode_to_parts(header)
    Base64.decode64(header).split(":")
  end

  def timestamp_nonce(diff = nil, scale = :minutes)
    ts = adjust(Time.now.utc, diff, scale)

    ts.strftime("%Y%m%dT%H%M%S.%6NZ")
  end

  def adjust(ts, diff, scale)
    return ts if diff.nil?

    case scale
    when :second, :seconds
      ts + diff
    when :minute, :minutes
      ts + (diff * 60)
    when :hour, :hours
      ts + (diff * 60 * 60)
    end
  end
end

File: elixir/support/support.ex
defmodule AppIdentity.Support do
  @moduledoc false

  def make_app(version, fuzz \\ nil) do
    version_fn =
      case version do
        1 -> :v1
        2 -> :v2
        3 -> :v3
        4 -> :v4
      end

    input = apply(__MODULE__, version_fn, [fuzz])

    AppIdentity.App.new(input)
  end

  def verified(%AppIdentity.App{} = app) do
    %{app | verified: true}
  end

  def verified(%{} = app) do
    verified(AppIdentity.App.new!(app))
  end

  def v1(fuzz \\ nil) do
    secret = random_hex(32)

    output = %{
      version: 1,
      id: uuidv4(),
      secret: fn -> secret end
    }

    if fuzz && is_integer(fuzz) && fuzz > 0 do
      Map.put_new(output, :config, %{"fuzz" => fuzz})
    else
      output
    end
  end

  def v2(fuzz \\ nil) do
    %{v1(fuzz) | version: 2}
  end

  def v3(fuzz \\ nil) do
    %{v1(fuzz) | version: 3}
  end

  def v4(fuzz \\ nil) do
    %{v1(fuzz) | version: 4}
  end

  def decode_to_parts(header) do
    header
    |> Base.url_decode64!(padding: false)
    |> String.split(":")
  end

  def build_padlock(app, options \\ []) do
    raw =
      Enum.join(
        [
          Keyword.get(options, :id) || app.id,
          Keyword.get(options, :nonce) || "nonce",
          padlock_secret(app, options)
        ],
        ":"
      )

    app
    |> padlock_algorithm(options)
    |> :crypto.hash(raw)
    |> Base.encode16(case: padlock_case(options))
  end

  def build_proof(app, padlock, options \\ []) do
    app_id = Keyword.get(options, :id) || app.id
    nonce = Keyword.get(options, :nonce) || "nonce"

    proof =
      case Keyword.get(options, :version, 1) do
        1 -> "#{app_id}:#{nonce}:#{padlock}"
        version -> "#{version}:#{app_id}:#{nonce}:#{padlock}"
      end

    Base.url_encode64(proof)
  end

  def timestamp_nonce(diff \\ nil, scale \\ :minutes) do
    "Etc/UTC"
    |> DateTime.now!()
    |> adjust_timestamp(diff, scale)
    |> DateTime.to_iso8601(:basic)
  end

  def adjust_timestamp(timestamp, nil, _) do
    timestamp
  end

  def adjust_timestamp(timestamp, diff, :seconds) do
    DateTime.add(timestamp, diff, :second)
  end

  def adjust_timestamp(timestamp, diff, :minutes) do
    DateTime.add(timestamp, diff * 60, :second)
  end

  def adjust_timestamp(timestamp, diff, :hours) do
    DateTime.add(timestamp, diff * 24 * 60, :second)
  end

  # The following code is adapted from https://hex.pm/packages/secure_random,
  # released under the Apache 2.0 license, copyright 2017 Patrick Robertson and
  # contributors.

  defp uuidv4 do
    <<u0::48, _::4, u1::12, _::2, u2::62>> = :crypto.strong_rand_bytes(16)
    <<g0::32, g1::16, g2::16, g3::16, g4::48>> = <<u0::48, 4::4, u1::12, 2::2, u2::62>>

    hex_pad(g0, 8) <>
      "-" <>
      hex_pad(g1, 4) <>
      "-" <>
      hex_pad(g2, 4) <>
      "-" <>
      hex_pad(g3, 4) <>
      "-" <>
      hex_pad(g4, 12)
  end

  defp random_hex(length) do
    length
    |> :crypto.strong_rand_bytes()
    |> Base.encode16(case: :lower)
  end

  defp hex_pad(hex, count) do
    hex = Integer.to_string(hex, 16)
    lower(hex, :binary.copy("0", count - byte_size(hex)))
  end

  defp lower(<<h, t::binary>>, acc) when h in ?A..?F, do: lower(t, acc <> <<h + 32>>)
  defp lower(<<h, t::binary>>, acc), do: lower(t, acc <> <<h>>)
  defp lower(<<>>, acc), do: acc

  defp padlock_secret(app, options) do
    secret = Keyword.get(options, :secret) || app.secret

    if is_function(secret, 0) do
      secret.()
    else
      secret
    end
  end

  defp padlock_algorithm(app, options) do
    case Keyword.get(options, :version) || app.version do
      1 -> :sha256
      2 -> :sha256
      3 -> :sha384
      4 -> :sha512
    end
  end

  defp padlock_case(options) do
    case Keyword.get(options, :case) || :random do
      :upper -> :upper
      :lower -> :lower
      :random -> if :rand.uniform(10) <= 5, do: :upper, else: :lower
    end
  end
end
